packages/plugin-ext/src/main/browser/plugin-authentication-service.ts

```diff
@@ -39,32 +37,35 @@ export class PluginAuthenticationServiceImpl extends AuthenticationServiceImpl i
         return super.login(id, scopes);
     }
 
-    private async tryActivateProvider(providerId: string): Promise<AuthenticationProvider> {
+    protected async tryActivateProvider(providerId: string): Promise<AuthenticationProvider> {
         this.pluginService.activateByEvent(getAuthenticationProviderActivationEvent(providerId));
 
-        let provider = this.authenticationProviders.get(providerId);
+        const provider = this.authenticationProviders.get(providerId);
         if (provider) {
             return provider;
         }
 
         // When activate has completed, the extension has made the call to `registerAuthenticationProvider`.
         // However, activate cannot block on this, so the renderer may not have gotten the event yet.
-        const didRegister: Promise<AuthenticationProvider> = new Promise((resolve, _) => {
-            this.onDidRegisterAuthenticationProvider(info => {
-                if (info.id === providerId) {
-                    provider = this.authenticationProviders.get(providerId);
-                    if (provider) {
-                        resolve(provider);
-                    } else {
-                        throw new Error(`No authentication provider '${providerId}' is currently registered.`);
-                    }
+        return Promise.race([
+            this.waitForProviderRegistration(providerId),
+            timeoutReject<AuthenticationProvider>(5000, 'Timed out waiting for authentication provider to register')
+        ]);
+    }
+
+    protected async waitForProviderRegistration(providerId: string): Promise<AuthenticationProvider> {
+        const waitForRegistration = new Deferred<AuthenticationProvider>();
+        const registration = this.onDidRegisterAuthenticationProvider(info => {
+            if (info.id === providerId) {
+                registration.dispose();
+                const provider = this.authenticationProviders.get(providerId);
+                if (provider) {
+                    waitForRegistration.resolve(provider);
+                } else {
+                    throw new Error(`No authentication provider '${providerId}' is currently registered.`);
```

Comment:
Suggestion: I don't believe we should throw this error here, but instead call `waitForRegistration.reject` with it? Looking at it, this seems to neither reject nor resolve the promise when no authentication provider with the given id is registered. Luckily, we always fail after 5 seconds using the timeout in that case, but I believe we should handle this more gracefully. (I know that this is kind of "old code")